photosynthesis_metrics/tv.py

```diff
@@ -0,0 +1,153 @@
+    r"""Interface of Total Variation.
+
+    Args:
+        x: Batch of images. Required to be 4D, channels first (N,C,H,W).
+        size_average: If size_average=True, total_variation of all images will be averaged as a scalar.
+        reduction_type: one of {'l1', 'l2', 'l2_squared'}
+
+    Returns:
+        Value of Structural Similarity (SSIM) index.
+
+    References:
+        .. [1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P.
+           (2004). Image quality assessment: From error visibility to
+           structural similarity. IEEE Transactions on Image Processing,
+           13, 600-612.
+           https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf,
+           :DOI:`10.1109/TIP.2003.819861`
+    """
+    _validate_input(x=x, y=x, kernel_size=3)
+    
+    tv = _compute_tv(x=x, 
+                     size_average=size_average, 
+                     reduction_type=reduction_type)
+
+    return tv
+
+
+
+class TVLoss(_Loss):
+    r"""Creates a criterion that measures the total variation error between
+    each element in the input :math:`x` and target :math:`y`.
+    See https://remi.flamary.com/demos/proxtv.html for reference.
+
+    If :attr:`reduction_type` set to ``'l2'`` loss can be described as:
+
+    .. math::
+        TV(x) = \sum_{N}\sqrt{\sum_{H, W, C}(|x_{:, :, i+1, j} - x_{:, :, i, j}|^2 +
+        |x_{:, :, i, j+1} - x_{:, :, i, j}|^2)}
+                                            
+    Else if :attr:`reduction_type` set to ``'l1'``:
+
+    .. math::
+        TV(x) = \sum_{N}\sum_{H, W, C}(|x_{:, :, i+1, j} - x_{:, :, i, j}| + 
+        |x_{:, :, i, j+1} - x_{:, :, i, j}|) $$
+
+    where :math:`N` is the batch size, `C` is the channel size.
+
+    .. math::
+        TVLoss(x, y) = |TV(x) - TV(y)|
+
+    :math:`x` and :math:`y` are tensors of arbitrary shapes with a total
+    of :math:`n` elements each.
+
+    Args:
+        size_average: If size_average=True, total_variation of all images will be averaged as a scalar.
+        reduction_type: one of {'l1', 'l2', 'l2_squared'}
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. Note: :attr:`size_average`
+            and :attr:`reduce` are in the process of being deprecated, and in the meantime,
+            specifying either of those two args will override :attr:`reduction`. Default: ``'mean'``
+    Examples::
+
+        >>> loss = SSIMLoss()
+        >>> prediction = torch.rand(3, 3, 256, 256, requires_grad=True)
+        >>> target = torch.rand(3, 3, 256, 256)
+        >>> output = loss(prediction, target, max_val=1.)
+        >>> output.backward()
+    """
+
+    def __init__(self, size_average: bool = True, reduction_type: str = 'l2', reduction: str = 'mean'):
+
```

Comment:
Formatting: remove blank line here.

From the Google style guide (section 3.5):
Two blank lines between top-level definitions, be they function or class definitions. One blank line between method definitions and between the class line and the first method. **No blank line following a def line**. Use single blank lines as you judge appropriate within functions or methods.